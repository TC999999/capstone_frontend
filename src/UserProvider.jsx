import UserContext from "./UserContext.js";
import { useState, useEffect } from "react";
import marketAPI from "../api.js";

const UserProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  //sets context for user for entire app
  const currentUser = (user) => {
    setUser(user);
  };

  // //On initial render, get user token from local storage, gets the current user from the api, and set as context
  useEffect(() => {
    async function getToken() {
      let token = localStorage.getItem("market-token");
      if (token) {
        let userInfo = await marketAPI.getCurrentUser();
        currentUser(userInfo);
      }
    }
    getToken();
  }, []);

  //These functions and user state will be used in context
  //Set the token in local storage, retrives the current user from the api, and sets the user as context on login
  const logIn = async (username, token) => {
    localStorage.setItem("market-token", token);
    let userInfo = await marketAPI.getUserInfo(username);
    currentUser(userInfo);
  };

  //remove token and user from local storage and set user context to null on logout
  const logOut = () => {
    localStorage.removeItem("market-token");
    currentUser(null);
  };

  //update current user in local storage and context
  const updateUser = async (username) => {
    let userInfo = await marketAPI.getUserInfo(username);
    currentUser(userInfo);
  };

  return (
    <UserContext.Provider
      value={{ user, currentUser, logIn, logOut, updateUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
