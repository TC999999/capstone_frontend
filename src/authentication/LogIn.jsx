import { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import marketAPI from "../../api";
import UserContext from "../UserContext";
import "../styles/Login.css";

const Login = ({ isLoading, setIsLoading }) => {
  const initialState = {
    username: "",
    password: "",
  };
  const { logIn } = useContext(UserContext);
  const navigate = useNavigate();
  const [err, setErr] = useState(false);
  const [message, setMessage] = useState("");
  const [formData, setFormData] = useState(initialState);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({ ...data, [name]: value }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    setErr(false);
    setMessage("");
    setIsLoading(true);
    const { username, password } = formData;
    try {
      const token = await marketAPI.logIn({
        username,
        password,
      });

      logIn(username, token);
      setFormData(initialState);
      navigate("/");
    } catch (err) {
      setErr(true);
      setMessage(err);
      setIsLoading(false);
    }
  };

  return (
    <div className="login-form">
      <h1>Log In</h1>
      <form onSubmit={handleSubmit}>
        <div className="username-div">
          <label htmlFor="username">Username: </label>
          <input
            id="login_username"
            type="text"
            name="username"
            placeholder="type your username here"
            value={formData.username}
            onChange={handleChange}
            required
          />
        </div>
        <div className="password-div">
          <label htmlFor="password">Password: </label>
          <input
            id="login_password"
            type="password"
            name="password"
            placeholder="type your password here"
            value={formData.password}
            onChange={handleChange}
            required
          />
        </div>
        {!isLoading ? (
          <div className="button-div">
            <button className="get-profile-button">Log In!</button>
          </div>
        ) : (
          <div>
            <p>Loading...</p>
          </div>
        )}
      </form>
      {err && <p className="err-msg">{message}</p>}
    </div>
  );
};

export default Login;
