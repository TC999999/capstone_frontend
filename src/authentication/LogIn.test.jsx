import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../UserProvider.jsx";
import LogIn from "./LogIn.jsx";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <LogIn />
      </UserProvider>
    </MemoryRouter>
  );
});

it("should match snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <LogIn />
      </UserProvider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

it("should have only username an password inputs", function () {
  const { container } = render(
    <MemoryRouter>
      <UserProvider>
        <LogIn />
      </UserProvider>
    </MemoryRouter>
  );

  const username_input = container.querySelector("#login_username");
  const password_input = container.querySelector("#login_password");
  const firstname_input = container.querySelector("#firstName");
  const lastname_input = container.querySelector("#lastName");

  expect(username_input).toBeInTheDocument();
  expect(password_input).toBeInTheDocument();
  expect(firstname_input).not.toBeInTheDocument();
  expect(lastname_input).not.toBeInTheDocument();
});
