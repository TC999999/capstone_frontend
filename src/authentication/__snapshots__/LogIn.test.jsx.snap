// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="login-form"
  >
    <h1>
      Log In
    </h1>
    <form>
      <div
        class="username-div"
      >
        <label
          for="username"
        >
          Username: 
        </label>
        <input
          id="login_username"
          name="username"
          placeholder="type your username here"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="password-div"
      >
        <label
          for="password"
        >
          Password: 
        </label>
        <input
          id="login_password"
          name="password"
          placeholder="type your password here"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="button-div"
      >
        <button
          class="get-profile-button"
        >
          Log In!
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
